### DISCUSSION TITLE
Make PersistRelations a public method

### DISCUSSION THREAD

**Discussion by patrickdemooij9** [umbraco/Umbraco-CMS#10552] [Features and ideas]
Current the method PersistRelations is a protected method: https://github.com/umbraco/Umbraco-CMS/blob/40dbfa7b8846c2835ddce08cdf690a9dc8719a94/src/Umbraco.Core/Persistence/Repositories/Implement/ContentRepositoryBase.cs#L964, but I think it would be better to make this public.

The PersistRelations function is now called whenever something updates or gets created. There is currently no way to rebuild the relations for multiple items. There might be a case where you are using a custom property picker that hasn't implemented the relations logic yet. However, if that logic does get implemented later on, you'll have to save every content item on your website again to make sure that those relations exists.

I've currently fixed this on my own website by copy pasting the PersistRelations logic in a different class, but it would be much easier if that was public. That way, people could also start writing rebuild relations logic as a backoffice button. Maybe we could even extend the Published Status dashboard with a Relations rebuild button.

**Comment by ronaldbarendse** [comment]
I think this makes sense and your use case is also a valid one, although it should then be exposed in the service layer (not the repository layer).

Looking at the code, it currently deletes all relations and then inserts them again, which is probably not something you want to do for all your unchanged content nodes (this will also unnecessarily defragment the database indexes BTW). This is currently not a major problem, as it's only executed for new or updated content and doesn't require fetching the existing relations, but we might want to look into that as well.
